eth/stagedsync/stage_bodies.go

```diff
@@ -252,15 +344,35 @@ Loop:
 			log.Trace("bodyLoop woken up by the incoming request")
 		}
 		d6 += time.Since(start)
+
+		return false, nil
 	}
-	if err := s.Update(tx, bodyProgress); err != nil {
-		return err
+
+	// kick off the loop and check for any reason to stop and break early
+	for !stopped {
+		shouldBreak, err := loopBody()
+		if err != nil {
+			return err
+		}
+		if shouldBreak {
+			break
+		}
 	}
+
+	// remove the temporary bucket for bodies stage
 	if !useExternalTx {
-		if err := tx.Commit(); err != nil {
+		bucketTx, err := cfg.db.BeginRw(context.Background())
+		if err != nil {
 			return err
 		}
+		defer bucketTx.Rollback()
+
+		bucketTx.ClearBucket("BodiesStage")
```

Comment:
what is max size of this bucket? why can't use ETL instead of bucket?

Reply:
Originally I was using an ETL but it meant that if I killed the erigon process the data would be gone when restarted.  Using the bucket means we can keep downloaded bodies there until they are ready for processing.  Requests quickly ramp up to around 32k in my goerli chain here and arrive in all kinds of strange orders so it made sense to hold on to them in the bucket.  The maximum size would be the amount of RAM consumed as part of the OOM issue, so quite large.

Reply:
Also, the existing code used a nice method of checking if there are any blocks to process, and could process them in batches to create the correct ordering on writing the bodies.  By using an ETL we'd need to load all of the data up front then process in one large batch rather than incrementally as data becomes available to write

Reply:
i'm fine with bucket - if we can limit it size